agent/consul/auto_encrypt_endpoint_test.go

```diff
@@ -0,0 +1,129 @@
+package consul
+
+import (
+	"crypto/x509"
+	"fmt"
+	"os"
+	"strings"
+	"testing"
+
+	"github.com/hashicorp/consul/agent/connect"
+	"github.com/hashicorp/consul/agent/structs"
+	"github.com/hashicorp/consul/testrpc"
+	"github.com/hashicorp/consul/tlsutil"
+	msgpackrpc "github.com/hashicorp/net-rpc-msgpackrpc"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestAutoEncryptSign(t *testing.T) {
+	t.Parallel()
+
+	type variant struct {
+		Config    tlsutil.Config
+		ConnError bool
+		RPCError  bool
+		Cert      string
+		Key       string
+	}
+	root := "../../test/ca/root.cer"
+	badRoot := "../../test/ca_path/cert1.crt"
+
+	variants := []variant{
```

Comment:
One improvement that could be made here is to add a subtests where the name of each subtest is something like `VerifyOutgoing fails`.

Here is a short read on table driven tests with subtests:
https://blog.golang.org/subtests

There's a few benefits to table driven subtests:
- You can run individual subtests on their own: `go test -run TestAutoEncryptSign/VerifyOutgoing_fails`
- The test case information is displayed in the test output, which makes it easier to line up what you expect vs what you got

Here's an example from envconsul where the name of the subtest is embedded in the test case struct:
https://github.com/hashicorp/envconsul/blob/master/runner_test.go#L181

This talk from Mitchell covers them as well:
https://www.youtube.com/watch?v=8hQG7QlcLBk

Reply:
Done!